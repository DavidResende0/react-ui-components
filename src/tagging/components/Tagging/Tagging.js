import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Row, Col } from 'patternfly-react';
import TagModifier from '../InnerComponents/TagModifier';
import TagView from '../InnerComponents/TagView';
import CategoryModifier from '../InnerComponents/CategoryModifier';
import ValueModifier from '../InnerComponents/ValueModifier';
import TaggingPropTypes from '../TaggingPropTypes';

class Tagging extends React.Component {
  onTagValueChange = (selectedTagValue) => {
    const action = {
      tagCategory: this.props.selectedTagCategory,
      tagValue: selectedTagValue,
    };

    // Single Tag Only
    if (this.props.options && this.props.options.onlySingleTag) {
      this.props.onSingleTagValueChange(action);
      return;
    }

    if (this.findSelectedTag(this.props.selectedTagCategory).singleValue) {
      this.props.onTagValueChange(action, this.props.options);
    } else {
      this.props.onTagMultiValueChange(action, this.props.options);
    }
  };

  onTagCategoryChange = selectedTagCategory =>
    this.props.onTagCategoryChange(selectedTagCategory);

  onTagDeleteClick = (tagCategory, tagValue) => {
    this.props.onTagDeleteClick({ tagCategory, tagValue }, this.props.options);
  };

  getvalues = () =>
    (this.findSelectedTag(this.props.selectedTagCategory) &&
      this.findSelectedTag(this.props.selectedTagCategory).values) ||
    [];

  findSelectedTag = (selectedTagCategory = { id: undefined }) =>
    this.props.tags.find(tag => tag.id === selectedTagCategory.id);

  isSelectedCategoryMultiValue = selectedTagCategory =>
    (this.findSelectedTag(selectedTagCategory.id) &&
      this.findSelectedTag(selectedTagCategory.id).singleValue) === true;

  tagCategories = this.props.tags.map(tag => ({
    description: tag.description,
    id: tag.id,
    singleValue: tag.singleValue,
  })) || [];

  render() {
    return (
      <Grid>
        <Row>
          <Col xs={12} md={8} lg={6}>
            <TagModifier hideHeader={this.props.options && this.props.options.hideHeaders}>
              <CategoryModifier
                selectedTagCategory={this.props.selectedTagCategory}
                onTagCategoryChange={this.props.onTagCategoryChange}
                tagCategories={this.tagCategories}
              />
              <ValueModifier
                onTagValueChange={this.onTagValueChange}
                selectedTagValue={this.props.selectedTagValue}
                multiValue={this.isSelectedCategoryMultiValue(this.props.selectedTagCategory)}
                values={this.getvalues()}
              />
            </TagModifier>
          </Col>
          <Col xs={12} md={4} lg={6}>
            <TagView
              hideHeader={this.props.options && this.props.options.hideHeaders}
              assignedTags={this.props.assignedTags}
              onTagDeleteClick={this.onTagDeleteClick}
            />
          </Col>
        </Row>
      </Grid>
    );
  }
}

Tagging.propTypes = {
  selectedTagCategory: TaggingPropTypes.category,
  selectedTagValue: TaggingPropTypes.value,
  tags: TaggingPropTypes.tags,
  assignedTags: TaggingPropTypes.tags,
  onTagDeleteClick: PropTypes.func.isRequired,
  onTagCategoryChange: PropTypes.func.isRequired,
  onTagValueChange: PropTypes.func.isRequired,
  onTagMultiValueChange: PropTypes.func.isRequired,
  onSingleTagValueChange: PropTypes.func.isRequired,
  options: PropTypes.shape({
    onlySingleTag: PropTypes.bool,
    hideHeaders: PropTypes.bool,
  }),
};

Tagging.defaultProps = {
  options: {
    onlySingleTag: false,
    hideHeaders: false,
  },
};

export default Tagging;
