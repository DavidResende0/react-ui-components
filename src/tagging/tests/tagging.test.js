import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tagging from '../components/tagging';


const tags = [
  { description: 'Name', id: 1, values: [{ description: 'Pepa', id: 11 }, { description: 'Franta', id: 12 }] },
  { description: 'Number', id: 2, values: [{ description: '1', id: 21 }, { description: '2', id: 22 }] },
  { description: 'Animal', id: 3, values: [{ description: 'Duck', id: 31 }, { description: 'Cat', id: 32 }, { description: 'Dog', id: 33 }] },
  { description: 'Food', id: 4, values: [{ description: 'Steak', id: 41 }, { description: 'Duck', id: 42 }, { description: 'Salad', id: 43 }] },
  {
    description: 'Something',
    id: 5,
    values: [{ description: 'Knedlik', id: 51 },
      { description: 'Daenerys Stormborn of the House Targaryen, First of Her Name,...and Mother of Dragons', id: 52 }],
  },
];

const selectedTagCategory = { description: 'animal', id: 1 };
const selectedTagValue = { description: 'duck', id: 1 };
const assignedTags = [{ tagCategory: { description: 'Name', id: 1 }, tagValues: [{ description: 'Pepa', id: 11 }] }];

const onChange = x => x;
const onDelete = x => x;


describe('Tagging component without redux mapping', () => {
  it('match snapshot', () => {
    const tree = mount(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onChange}
      onTagCategoryChange={onChange}
      onTagDeleteClick={onDelete}
      selectedTagCategory={selectedTagCategory}
      selectedTagValue={selectedTagValue}
    />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should call methods', () => {
    const onTagCategoryChange = jest.fn();
    const onTagValueChange = jest.fn();
    const onTagDeleteClick = jest.fn();
    const wrapper = shallow(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onTagValueChange}
      onTagCategoryChange={onTagCategoryChange}
      onTagDeleteClick={onTagDeleteClick}
      selectedTagCategory={selectedTagCategory}
      selectedTagValue={selectedTagValue}
    />);
    wrapper.instance().onTagCategoryChange('xaxa');
    expect(onTagCategoryChange.mock.calls.length).toEqual(1);
    wrapper.instance().onTagValueChange('wawa');
    expect(onTagCategoryChange.mock.calls.length).toEqual(1);
    wrapper.instance().onTagDeleteClick('wowo');
    expect(onTagCategoryChange.mock.calls.length).toEqual(1);
  });
});
