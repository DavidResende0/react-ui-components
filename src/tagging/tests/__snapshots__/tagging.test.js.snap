// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tagging component without redux mapping match snapshot 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={true}
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <Col
      bsClass="col"
      componentClass="div"
      lg={6}
      md={8}
      xs={12}
    >
      <TagModifier
        header="Add/Modify tag"
        hideHeader={false}
      >
        <CategoryModifier
          categoryLabel="Category"
          onTagCategoryChange={[MockFunction]}
          selectedTagCategory={
            Object {
              "description": "animal",
              "id": 1,
            }
          }
          tagCategories={
            Array [
              Object {
                "description": "Name",
                "id": 1,
                "singleValue": undefined,
              },
              Object {
                "description": "Number",
                "id": 2,
                "singleValue": undefined,
              },
              Object {
                "description": "Animal",
                "id": 3,
                "singleValue": undefined,
              },
              Object {
                "description": "Food",
                "id": 4,
                "singleValue": undefined,
              },
              Object {
                "description": "Something",
                "id": 5,
                "singleValue": undefined,
              },
            ]
          }
        />
        <ValueModifier
          multiValue={false}
          onTagValueChange={[Function]}
          selectedTagValues={
            Array [
              Object {
                "description": "Pepa",
                "id": 11,
              },
            ]
          }
          valueLabel="Value"
          values={
            Array [
              Object {
                "description": "Pepa",
                "id": 11,
              },
              Object {
                "description": "Franta",
                "id": 12,
              },
            ]
          }
        />
      </TagModifier>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      lg={6}
      md={4}
      xs={12}
    >
      <TagView
        assignedTags={
          Array [
            Object {
              "description": "Name",
              "id": 1,
              "values": Array [
                Object {
                  "description": "Pepa",
                  "id": 11,
                },
              ],
            },
          ]
        }
        header="Assigned tags"
        hideHeader={false}
        onTagDeleteClick={[Function]}
      />
    </Col>
  </Row>
</Grid>
`;
